### PR TITLE
Preserve original lifetime when performing pin-projection

### PR DESCRIPTION
**Pull request by Aaron1011**
As described in issue https://github.com/taiki-e/pin-project/issues/65, the `project` method currently loses lifetime information.

Specifically, given a `&'a mut Pin<&'pin mut T>`, we would like `project` to return a `ProjectionType<'pin>` - that is, a type with a lifetime tied to the lifetime of the inside of the original `Pin`.

However, `project` currently returns a `ProjectionType<'a>` - that is, a type with a lifetime tied to the lifetime of the 'outside' of the `Pin`. This means that pin-projected types will not live as long as the inside of the `Pin` they were derived from - which prevents returning fields from methods.

This PR lifts this limitation, by ensuring that the returning projection struct/enum has a lifetime of `'pin` - that is, the lifetime in the `Pin<&'pin mut T>` that the projection was performed on.

Most of this PR is just boilerplate to make things work on arbitrary structs and enums. The important part is the additional `unsafe` code added. Ultimately, it relies on this code being valid:

```rust
use std::ops::DerefMut;

fn upgrade_deref_mut<'a, T>(mut val: &'a mut T) -> &'a mut T {
    let my_ref: &mut T = val.deref_mut();
    let transmuted: &'a mut T = unsafe { core::mem::transmute(my_ref) };
    transmuted
}
```

That is, we need to be able to 'upgrade' the lifetime of the mutable reference returned from the `<&mut T as DerefMut>::deref_mut`. I believe this is sound - `DerefMut` is essentially a no-op for mutable references. We're only adjusting the compiler's view of the lifetime, not performing any raw pointer tricks that might make Stacked Borrows angry.

That being said, I'm not 100% sure that this is correct.

@RalfJung: Would you be able to take a look at this?

Note that this only relies on the built-in `DerefMut` impl for `&mut T`, not any user-provided `&mut DerefMut` impls. I think that this kind of 'upgrading' logic could potentially be extended to user-provided types that fulfill the requirement of `StableDeref` - that is, the mutable references they return live for as long as the original type. However, pin-project doesn't need to care about this.

### CODE REVIEW

**Comment by Aaron1011** [comment]
Assuming that this is indeed sound, it might be useful for Rust to provide this kind of upgrading behavior on `Pin` itself.

Making it would for arbitrary user types would require introducing a weaker form of `StableDeref` into the compiler (it only needs to allow transmuting mutable references, not any of the raw pointer operations that have given us trouble).  However, it could be added only for `Pin<&mut T>`:

```rust
impl<'a, T> Pin<&'a mut T> {
	fn as_mut_better(&mut self) -> Pin<'a mut T> {
        /// The impl from this crate
    }
}
```

**Comment by Aaron1011** [comment]
Thinking about this some more: I believe the `transmute` is essentially equivalent to this trait and impl:

```rust
use std::ops::DerefMut;

trait BetterDeref<'a> {
    type Target: ?Sized;
    fn better_mut(&'a mut self) -> &'a mut Self::Target;
}

impl<'a, T: ?Sized> BetterDeref<'a> for &'a mut T {
    type Target = T;
    fn better_mut(&'a mut self) -> &'a mut Self::Target {
        self
    }
}
```

If the two are equivalent, then the transmute must be sound, since this trait is imlpemented in purely safe code.

File: pin-project-internal/src/pin_project/mod.rs [hunk 1]
```diff
@@ -125,6 +138,21 @@ impl Context {
         generics
     }
 
+    fn proj_trait_generics(&self) -> ProjTraitGenerics<'_> {
+        //let mut orig_generics_for_impl = orig.clone();
+        //crate::utils::proj_generics(&mut orig_generics_for_impl, self.lifetime.clone());
+        //let (impl_generics, modified_ty_generics, _) = orig_generics_for_impl.split_for_impl();
```
Comment (taiki-e) [inline pin-project-internal/src/pin_project/mod.rs]: Please remove the above 3 lines.

File: pin-project-internal/src/pin_project/mod.rs [hunk 2]
```diff
@@ -343,11 +371,15 @@ impl Context {
         let proj_generics = self.proj_generics();
         let proj_ty_generics = proj_generics.split_for_impl().1;
 
-        let (orig_generics, _, orig_where_clause) = self.generics.split_for_impl();
+        let mut trait_generics = self.generics.clone();
+        crate::utils::proj_generics(&mut trait_generics, lifetime.clone());
+        //trait_generics.params.push(GenericParam::Lifetime(LifetimeDef::new(lifetime.clone())));
```
Comment (taiki-e) [inline pin-project-internal/src/pin_project/mod.rs]:
```suggestion
```

File: pin-project-internal/src/pin_project/structs.rs [hunk 3]
```diff
@@ -31,17 +31,29 @@ pub(super) fn parse(cx: &mut Context, mut item: ItemStruct) -> Result<TokenStrea
     let Context { proj_ident, proj_trait, orig_ident, lifetime, .. } = &cx;
     let proj_generics = cx.proj_generics();
     let proj_ty_generics = proj_generics.split_for_impl().1;
-    let (impl_generics, ty_generics, where_clause) = item.generics.split_for_impl();
+
+    /*let mut orig_generics_for_impl = item.generics.clone();
+    crate::utils::proj_generics(&mut orig_generics_for_impl, lifetime.clone());
+    let (impl_generics, modified_ty_generics, _) = orig_generics_for_impl.split_for_impl();
+
+    let (_, ty_generics, where_clause) = item.generics.split_for_impl();*/
```
Comment (taiki-e) [inline pin-project-internal/src/pin_project/structs.rs]: Please remove the above 5 lines.

File: src/lib.rs [hunk 4]
```diff
@@ -161,6 +161,78 @@ pub use pin_project_internal::project;
...
+        // is the same as 'a. In order words, calling deref_mut on a mutable reference
+        // is a no-op.
+        let as_mut: Pin<&mut T> = Pin::as_mut(self);
+
+        // Next, we unwrap the Pin<&mut T> by calling 'Pin::get_unchecked_mut'.
+        // This gives us access to the underlying mutable reference
+        #[allow(unsafe_code)]
+        let raw_mut: &mut T = unsafe { Pin::get_unchecked_mut(as_mut) };
+
+        // NExt, we transmute the mutable reference, changing its lifetime to 'a.
```
Comment (taiki-e) [inline src/lib.rs]:
```suggestion
        // Next, we transmute the mutable reference, changing its lifetime to 'a.
```

File: src/lib.rs [hunk 5]
```diff
@@ -161,6 +161,78 @@ pub use pin_project_internal::project;
...
+/// with a `&mut Pin<'pin &mut T>`, the 'pin lifetime is lost.
+///
+/// This can cause issues when trying to return the result of a pin projection
+/// from a method - e.g. `fn foo<'a>(mut self: Pin<&'a mut Self>) -> Pin<&'a mut T>`
+///
+/// The `ProjectThrough` trait was introduced to solve this issue.
+/// Normally, you'll never need to interact with this type
+/// directly - it's used internally by pin-project to allow expressing
+/// the proper lifetime.
+pub trait ProjectThrough<'a> {
```
Comment (taiki-e) [inline src/lib.rs]: Maybe this trait should be sealed?
```rust
mod private_project_through {
    use super::*;

    pub trait Sealed {}

    impl<T> Sealed for Pin<&mut T> {}
}

pub trait ProjectThrough<'a>: private_project_through::Sealed {
    // ...
}
```

File: pin-project-internal/src/pin_project/enums.rs [hunk 6]
```diff
@@ -34,17 +34,23 @@ pub(super) fn parse(cx: &mut Context, mut item: ItemEnum) -> Result<TokenStream>
...
         enum #proj_ident #proj_generics #where_clause { #(#proj_variants,)* }
     };
+
+    let crate_path = crate::utils::crate_path();
+
     proj_items.extend(quote! {
-        impl #impl_generics #proj_trait #ty_generics for ::core::pin::Pin<&mut #orig_ident #ty_generics> #where_clause {
-            fn project<#lifetime>(&#lifetime mut self) -> #proj_ident #proj_ty_generics #where_clause {
+        impl #impl_generics #proj_trait #ty_generics for ::core::pin::Pin<&#lifetime mut #orig_ident #orig_ty_generics> #where_clause {
+            fn project(&mut self) -> #proj_ident #proj_ty_generics #where_clause {
```
Comment (taiki-e) [inline pin-project-internal/src/pin_project/enums.rs]:
```suggestion
            fn project(&mut self) -> #proj_ident #proj_ty_generics {
```
Preexisting, but as it is already declared in the impl block, it looks unnecessary.

File: pin-project-internal/src/pin_project/mod.rs [hunk 7]
```diff
@@ -343,11 +371,15 @@ impl Context {
...
+
+        let (orig_generics, _, orig_where_clause) = trait_generics.split_for_impl();
 
         quote! {
             trait #proj_trait #orig_generics {
-                fn project<#lifetime>(&#lifetime mut self) -> #proj_ident #proj_ty_generics #orig_where_clause;
+                fn project(&mut self) -> #proj_ident #proj_ty_generics #orig_where_clause;
```
Comment (taiki-e) [inline pin-project-internal/src/pin_project/mod.rs]:
```suggestion
                fn project(&mut self) -> #proj_ident #proj_ty_generics;
```

File: pin-project-internal/src/pin_project/structs.rs [hunk 8]
```diff
@@ -31,17 +31,29 @@ pub(super) fn parse(cx: &mut Context, mut item: ItemStruct) -> Result<TokenStrea
...
         struct #proj_ident #proj_generics #where_clause #proj_fields
     };
+
+    let crate_path = crate::utils::crate_path();
+
     proj_items.extend(quote! {
-        impl #impl_generics #proj_trait #ty_generics for ::core::pin::Pin<&mut #orig_ident #ty_generics> #where_clause {
-            fn project<#lifetime>(&#lifetime mut self) -> #proj_ident #proj_ty_generics #where_clause {
+        impl #impl_generics #proj_trait #ty_generics for ::core::pin::Pin<&#lifetime mut #orig_ident #orig_ty_generics> #where_clause {
+            fn project(&mut self) -> #proj_ident #proj_ty_generics #where_clause {
```
Comment (taiki-e) [inline pin-project-internal/src/pin_project/structs.rs]:
```suggestion
            fn project(&mut self) -> #proj_ident #proj_ty_generics {
```

File: src/lib.rs [hunk 9]
```diff
@@ -161,6 +161,78 @@ pub use pin_project_internal::project;
 #[allow(unsafe_code)]
 pub unsafe trait UnsafeUnpin {}
 
+use core::pin::Pin;
+
+/// A helper trait to allow projecting through a mutable reference,
+/// while preserving lifetimes.
+///
+/// Normally, `Pin::as_mut` can be used to convert a '&mut Pin' to a owned 'Pin'
```
Comment (taiki-e) [inline src/lib.rs]:
```suggestion
/// Normally, `Pin::as_mut` can be used to convert a '&mut Pin' to an owned 'Pin'
```

File: src/lib.rs [hunk 10]
```diff
@@ -161,6 +161,78 @@ pub use pin_project_internal::project;
...
+/// However, since `Pin::as_mut'`needs to work with generic `DerefMut` impls,
```
Comment (taiki-e) [inline src/lib.rs]:
```suggestion
/// However, since `Pin::as_mut'` needs to work with generic `DerefMut` impls,
```

File: src/lib.rs [hunk 11]
```diff
@@ -161,6 +161,78 @@ pub use pin_project_internal::project;
...
+        // This method is fairly tricky. We start out with
+        // a `&'p mut Pin<&'a mut T>`, which we want
+        // to convert into a `Pin<&'a mut T>`
+        //
+        // Effectively, we want to discard the outer '&mut'
+        // reference, and just work with the inner 'Pin'
+        // We accomplish this in four steps:
+        //
+        // First, call Pin::as_mut(self). This
+        // converts our '&'p mut Pin<&'a mut T>` to
```
Comment (taiki-e) [inline src/lib.rs]:
```suggestion
        // converts our `&'p mut Pin<&'a mut T>` to
```

File: src/lib.rs [hunk 12]
```diff
@@ -161,6 +161,78 @@ pub use pin_project_internal::project;
...
+        // This gives us access to the underlying mutable reference
+        #[allow(unsafe_code)]
+        let raw_mut: &mut T = unsafe { Pin::get_unchecked_mut(as_mut) };
+
+        // NExt, we transmute the mutable reference, changing its lifetime to 'a.
+        // Here, we rely on the behavior of DerefMut for mutable references
+        // described above.
+        //
+        // This is the core of this method - everything else is just to deconstruct
+        // and reconstrut a Pin
```
Comment (taiki-e) [inline src/lib.rs]:
```suggestion
        // and reconstruct a Pin.
```

**Comment by Aaron1011** [comment]
It turns out that this is actually unsound - the output lifetime isn't tied to the input lifetime, so it's possible to creating aliasing mutable references by calling `project` twice.

I'm investigating if this can be made sound.

**Comment by Aaron1011** [comment]
Unfortunately, I think it's impossible to do this soundly in the `project` method.

There are two conflicting requirements:
* The projection type returned from `Project` should **not** be required to outlive `'a` in `Pin<&'a mut MyType>`. This allows us to call `project()` multiple times, and to use the original `Pin<&'a mut MyType>` after calling `project()`.
* THe projection type returned from `Project` **should** be required to outlive `'a` in `Pin<&'a mut MyType>`. This allow us to write code like:

```rust
fn get_pin_mut<'a>(mut self: Pin<&'a mut Self>) -> Pin<&'a mut T> {
    self.project().pinned
}
```

where the returned type outlives the `self` parameter passed into the method.

It is impossible for one lifetime that both outlives and does not outlive another lifetime.

Instead, I think we need to add a new method: `project_into`. This method will take `self` by value (effectively adding the pre-https://github.com/taiki-e/pin-project/pull/47 behavior back). Users will choose between `project` or `project_into` based on their needs.

**Comment by Aaron1011** [comment]
Unfortunately, I think it's impossible to do this soundly in the `project` method.

There are two conflicting requirements:
* The projection type returned from `Project` should **not** be required to outlive `'a` in `Pin<&'a mut MyType>`. This allows us to call `project()` multiple times, and to use the original `Pin<&'a mut MyType>` after calling `project()`.
* THe projection type returned from `Project` **should** be required to outlive `'a` in `Pin<&'a mut MyType>`. This allow us to write code like:

```rust
fn get_pin_mut<'a>(mut self: Pin<&'a mut Self>) -> Pin<&'a mut T> {
    self.project().pinned
}
```

where the returned type outlives the `self` parameter passed into the method.

It is impossible for one lifetime that both outlives and does not outlive another lifetime.

Instead, I think we need to add a new method: `project_into`. This method will take `self` by value (effectively adding the pre-https://github.com/taiki-e/pin-project/pull/47 behavior back). Users will choose between `project` or `project_into` based on their needs.

**Comment by Aaron1011** [comment]
Superceded by https://github.com/taiki-e/pin-project/pull/69

**Comment by RalfJung** [comment]
I am very confused by this PR, this seems entirely unnecessary -- why don't you reborrow the pinned reference? See what I wrote at https://github.com/taiki-e/pin-project/pull/47#issuecomment-530467225.

**Pull request closed without merging**